### ISSUE TITLE
[Feature] Conan 2 cache concurrency

### ISSUE THREAD

**Issue by memsharded** [conan-io/conan#15840]
This ticket is to gather and centralize all the related tickets.
There are 3 different aspects to make the cache concurrent:
- ``conan config install`` and other commands that can change the Conan home configuration concurrently
- Package installation and building concurrently by different processes
- Better management of "dirty" status to recover gently from crashes and interrupted operations (already a feature, but requires refactoring and alignment with the concurrency)

Out of the scope:
- Concurrency at multi-machine or multi-OS level. The planned supported concurrency is at the machine level, from the same OS. OS sync mechanism file-locks will be used, and that is only guaranteed to work at the OS level, no shared or mounted drives
- Sharing the same cache among different Conan versions. The same Conan version must be used for the same cache.

Conan config install concurrent:
- https://github.com/conan-io/conan/issues/6233
- https://github.com/conan-io/conan/issues/15018

Package concurrency
- https://github.com/conan-io/conan/issues/14570
- https://github.com/conan-io/conan/issues/11033
- https://github.com/conan-io/conan/issues/8241
- https://github.com/conan-io/conan/issues/6753
- https://github.com/conan-io/conan/issues/4648
- https://github.com/conan-io/conan/issues/5505
- https://github.com/conan-io/conan/issues/7181

**Comment by darakelian** [comment]
I was told to comment here for posterity:
My team has a large repo that consists of a "core" package and hundreds of components that are being packaged up as individual conan packages (with a dependency on the "core" package.) Previously in conan1, our build script that called `conan create` in parallel for all of the component packages worked flawlessly (build times of around 3 minutes on our CI boxes) however upon switching to conan2 I noticed that often the `conan create` command would error due to the cache concurrency issues. To solve this I have had to run our script in serial mode which massively inflates our build time to approximately 25 minutes (these packages are header-only so in theory we are IO-bound anyways but it appears we are still able to massively benefit from parallelization.) Given that this behavior of building multiple conan packages in parallel seemed to work in conan1, it would be great if the cache concurrency could be improved here in conan2.

**Comment by memsharded** [comment]
Thanks for the feedback @darakelian, this is something that is planned, but it might take a little bit.

In the meantime, you might want to try other approaches to speed up things. For example, you can run things that can run in parallel in different ``CONAN_HOME`` folders, then accumulate packages in the same one with ``conan cache save/restore``. Avoiding extra downloads can be done by sharing the "download cache folder", which is concurrent and can be shared among other parallel Conan home folders. For building a dependency graph in parallel, the ``conan graph build-order`` will give a list of lists of things that can be safely build in parallel too.

**Comment by nextsilicon-itay-bookstein** [comment]
Since the build phases themselves are often already parallel, to me it seems that the main utility of this is to parallelize across the configure steps of all packages, which are always completely serial and therefore under-utilize the machine. I'm also interested, wanting to build a not-insignificant matrix of packages * profiles * build_type.

**Comment by ViliusSutkus89** [comment]
Hello,

Not opening a new issue, since I assume the best place for my bug is this issue.

Parallel `conan install`'s are always failing on an empty Conan package cache. Steps to reproduce:
```sh
rm -r ~/.conan2/p
conan install . --output-folder=build-a &
conan install . --output-folder=build-b &
```

I get errors like:
```
ERROR: Package 'zlib/1.3.1' not resolved: Reference 'zlib/1.3.1#f52e03ae3d251dec704634230cd806a2%1708593606.497' already exists.
```

I've set up a [small test case](https://github.com/opendocument-app/ConanAndroidGradlePlugin/blob/issue_1/tests/parallel_conan_install/testcase.py) to illustrate. Here's the [CI run](https://github.com/opendocument-app/ConanAndroidGradlePlugin/actions/runs/10311546628/job/28545329963).

I care about this because in my workflow I need to run four conan installs - one for each of Android's architectures. As a workaround, I run one of them, wait for it to finish and then run the three remaining in parallel. It just takes longer to complete. Since I have a workaround, this would be more of a performance improvement, than a correctness issue.

**Comment by memsharded** [comment]
Thanks for the feedback @ViliusSutkus89, good in this thread, yes. 

Another possible workaround is to do something like fetch first the recipe (something like one ``conan graph info ...``), which will be faster than having to wait for 1 configuration, then launch the different builds for the different architectures in parallel. As the race condition seems to be happening in the download of the ``zlib`` recipe, this will likely work better and be faster.

**Comment by ViliusSutkus89** [comment]
Tried graph generation, It could work, but there's a caveat that different builds can't share any dependencies between them. Previously the error was :
```
ERROR: Package 'zlib/1.3.1' not resolved: Reference 'zlib/1.3.1#f52e03ae3d251dec704634230cd806a2%1708593606.497' already exists
```

With the graph pre-generated and with the revisions known, the concurrent write error is triggered a bit later:
```
ERROR: Reference 'zlib/1.3.1#f52e03ae3d251dec704634230cd806a2%1708593606.497:b647c43bfefae3f830561ca202b6cfd935b56205#6b307bbcbae23635c4006543ffdbf3ef%1708593932.513' already exists
```

In my test case both builds use the same zlib, because both builds are actually from the same arch. But I've checked and this also happens with different profiles that share the same arch agnostic dependency.

**Comment by memsharded** [comment]
I see, yes, the ``conan graph info`` approach helps in the case of parallel building of the same graph, but there might still be issues with other different parallel jobs. If sharing the same cache, all the ``conan graph info`` + failed ``conan install`` for all profiles should be launched first sequentially before launching the parallelism. But it seems this would result in a "dirtier" pipeline, so it sounds your original approach would be better here then.

**Comment by marlamb** [comment]
Hi, I noticed #11480 , but would like to do something similar, but slightly different and am now wondering if this is a good idea.
In my setup I would like to export many *different* recipes in parallel. In my first local tests I did not see any issues. But as concurrency problems tend to turn up in a non-reproducible manner, I would like to know if something like this (pseudo code) is officially possible without potential concurrency issues (latest `conan` version):
```
for recipe_path in set(recipe_paths) do in parallel:
    subprocess.check_call(["conan", "export", recipe_path])
```
Thanks in advance!

**Comment by memsharded** [comment]
Hi @marlamb, https://github.com/conan-io/conan/issues/11480 was pre Conan 2.0

With the Conan2 new PythonAPI, exporting recipes is very fast, might not even need concurrency at all. I'd recommend to do a custom command that iterates and calls the export API. Please try that and open new tickets for any further question about it.

**Comment by redradist** [comment]
> Hi @marlamb, #11480 was pre Conan 2.0
> 
> With the Conan2 new PythonAPI, exporting recipes is very fast, might not even need concurrency at all. I'd recommend to do a custom command that iterates and calls the export API. Please try that and open new tickets for any further question about it.

@memsharded Tools that rely on fast execution to prevent concurrency issues, would suffer from random tickets create for it ...
Concurrency is not an option, it definitely should be implemented !!

**Comment by marlamb** [comment]
At least for exporting recipes I could confirm that using the API is indeed sufficiently quick compared to running parallel subprocesses. On my local machine (admittedly no proof) using the API on a single thread was twice as fast as using the CLI via subprocess calls utilizing 16 cores on a medium number of recipes (roughly 500). In absolute numbers it is about few seconds, which looks to me very fast compared to other actions that might be interesting on that number of recipes (e.g. creating lockfiles).

Apart from that I agree that I would naturally expect the cache to be concurrency safe.

**Comment by memsharded** [comment]
Thanks for the feedback @marlamb 

Indeed, we are not opposed or anything like that to make the cache concurrent, it is simply that it is a quite challenging feature, and we haven't had the time to prioritize yet compared with other more functional features like workspaces). Hopefully we will be able to approach it in first half of 2025.

Still, the point that using the API can be order of magnitude faster than using the CLI in some scenarios, like the multiple recipe export, must be taken into account, it is not just a workaround to not being able to launch the Conan client in parallel, sometimes it is the best solution to a problem.

**Comment by redradist** [comment]
> I see, yes, the `conan graph info` approach helps in the case of parallel building of the same graph, but there might still be issues with other different parallel jobs. If sharing the same cache, all the `conan graph info` + failed `conan install` for all profiles should be launched first sequentially before launching the parallelism. But it seems this would result in a "dirtier" pipeline, so it sounds your original approach would be better here then.

@memsharded For me this approach does not work for 2 different build types, for Debug and Release.
When I try to do under lock `conan graph info` then create a package and then again under lock object `conan graph info` then create a package ... and it still broken

**Comment by memsharded** [comment]
Sorry @redradist it is not very clear what you mean with your issue, what would be "under lock" and what would be a detailed sequence of commands, and which of those commands are executed in parallel and creating concurrency issues, and what are exactly the error messages that produce.

**Comment by redradist** [comment]
> Sorry @redradist it is not very clear what you mean with your issue, what would be "under lock" and what would be a detailed sequence of commands, and which of those commands are executed in parallel and creating concurrency issues, and what are exactly the error messages that produce.

@memsharded I mean that I've created something like this:
```python
def run_command(...):
    ...

with filelock_for_specific_package:
    run_command("conan graph info")
    run_command("conan create ...")
```

**Comment by memsharded** [comment]
``conan graph info`` and ``conan create`` do install and potentially build dependencies. So locking for 1 specific package will not protect against possible race conditions in the dependencies being downloaded/built concurrently in the cache.

**Comment by redradist** [comment]
> `conan graph info` and `conan create` do install and potentially build dependencies. So locking for 1 specific package will not protect against possible race conditions in the dependencies being downloaded/built concurrently in the cache.

Yes, agree, because this filelock should be applied not outside of conan, but inside for each package and its dependecies as well.
It will prevent race conditions

**Comment by beabanyai** [comment]
Hi,

We are setting up a common storage path for our application packages, in order to spare storage on a local machine, which can run parallel builds.
Can this issue impact the concurrent usage (only read actions) of a package from the same storage_path?

Approximately when will the cache concurrency be safe to use/ when do you plan to release it?
Thanks!

**Comment by AbrilRBS** [comment]
> Can this issue impact the concurrent usage (only read actions) of a package from the same storage_path

While [our guideline still stands](https://docs.conan.io/2/knowledge/guidelines.html) that in general no parallaleism is supported, **pure read-only** concurrent access is more than likely to work - the main limitation there being concurrent db access (for which sqlite has some support).

But as soon as one of your parallel builds modifies the cache, you risk triggering the issues that have been linked to this issue over time, and so different approaches might be needed. (For example, for CI systems in the same shared drive, adding a "file download cache" with the `core.download:download_cache` conf, that can be more than enough, etc.

> Approximately when will the cache concurrency be safe to use/ when do you plan to release it?

There is ongoing work for this, but currently we have no concrete timeline for this. When progress is made, we'll update this issue accordingly :)

**Comment by redradist** [comment]
> > Can this issue impact the concurrent usage (only read actions) of a package from the same storage_path
> 
> While [our guideline still stands](https://docs.conan.io/2/knowledge/guidelines.html) that in general no parallaleism is supported, **pure read-only** concurrent access is more than likely to work - the main limitation there being concurrent db access (for which sqlite has some support).
> 
> But as soon as one of your parallel builds modifies the cache, you risk triggering the issues that have been linked to this issue over time, and so different approaches might be needed. (For example, for CI systems in the same shared drive, adding a "file download cache" with the `core.download:download_cache` conf, that can be more than enough, etc.
> 
> > Approximately when will the cache concurrency be safe to use/ when do you plan to release it?
> 
> There is ongoing work for this, but currently we have no concrete timeline for this. When progress is made, we'll update this issue accordingly :)

@AbrilRBS Check [conanex](https://github.com/redradist/conanex), there I have added '--global-lock' to preform using 'conan' parallel operations

**Comment by bnason-nf** [comment]
Hi,

I have set of changes that to allow Conan to work from multiple simultaneous processes without obvious cache conflicts and so on. Full disclosure: since I'm not familiar with the Conan code base, the changes were made with the assistance of an LLM, but I have manually reviewed the code and it seems correct and reasonable quality  (i.e., it doesn't seem to be slop). The change includes some unit tests, but I'm sure for something of this scope it would need a lot more testing.

If I were to submit a pull request, is that something you'd be interested in? Aside from the steps in CONTRIBUTING.md, is there anything else that you would like to see?

Thanks,
Benbuck

**Comment by memsharded** [comment]
Hi @bnason-nf 

Thanks for your feedback and trying to implement some concurrency support.
We are at the moment investigating how this could be moved forward, and thinking about a bit more deep re-designs of the cache that could allow some lock-free sharing (as real multi-OS distributed locking is almost impossible without a server/daemon or a full postgres DB, as filelocks are not cross-platform, and consequently sqlite3 cannot be used that way either).

But I would be interested in any case on what the AI did in this regard, having a look, etc, even if the PR can't be moved forward as-is, and we might close it, but if you already have the code, sure, please open it, it will cost nothing :)

Many thanks!

**Comment by bnason-nf** [comment]
Hi @memsharded,

Thanks, I'm doing some more fixes and testing, and I hope to submit a PR some next week or so. To your point, the PR assumes parallel processes on a single machine, which seems like a good improvement over the current status, but is certainly not going to solve anything for a distributed environment if that's a requirement.

-Benbuck

**Comment by bnason-nf** [comment]
Hi @memsharded,

I just created PR #19555. This is a very large change, so I'm positive that there will be many things in it that aren't what you would want. However, I think the big picture question is whether having multi-process concurrency is something you'd want (maybe before having full distributed concurrency). If so, then I'm happy to discuss whatever you'd like to see changed in order to get this landed. In any case, thanks in advance for taking a look at this, I appreciate your time.

-Benbuck

**Comment by Eugene-Alexeev** [comment]
how far are we from conan installing and building dependencies in parallel? even approximately, just to plan some internal things at my project

**Comment by memsharded** [comment]
Hi @Eugene-Alexeev 

Quick question, are you being hit by https://github.com/conan-io/conan/issues/19181?
Because if that is not the case, Conan can already install pre-compiled binaries from servers concurrently, with the number of threads defined by the ``conf`` named ``core.download:parallel``.

In any case, it doesn't seem this will happen any time soon, because it is simply too complex, for example our recent efforts have been blocked by  Windows antivirus programs doing locking on some files, so even a simple ``os.replace`` can not be done atomically. With the current constraints, having a filesystem based Conan cache that can be shared for concurrency is almost an unsolvable problem. We will keep investigating, but in any case it will take time.

**Comment by mneumei** [comment]
This would be a huge benefit for large monorepo projects.

We try to export and build 500+ packages in parallel as we have dependencies between the packages. This also makes it hard to build on multiple machines (or in mutiple different caches). Having proper handling of the cache would speed things up a lot.
